### PR TITLE
Pulled sql operator return type inference out

### PR DESCRIPTION
**Pull request by lziq**
*Major changes in this PR:*
1. Pulled sql operator return type inference out of `StaticTypeInferenceVisitorTransform.kt` file, and migrate them to the standard plugin package. 

*Minor changes in this PR:*
1. Added `ScalarTypeSystem` as a configuration option to `CompilerPipeline`. 
2. Renamed the package `org.partiql.lang.ots` to `org.partiql.lang.ots_work` 
3. Some minor refactoring in `StaticTypeInferenceVisitorTransform.kt`, e.g. renaming variable `nary` to `processedNode`.

### CODE REVIEW

**Review by vgapeyev — COMMENTED** [review]
Ok, this review is not comprehensive in coverage, but I reviewed a few spots deeply enough to offer some meaningful directional feedback.  It should apply to other parts in this PR, as well as to the earlier work. 

I am growing convinced that, to help assure correctness in the OTS project, as well as to maintain our sanity, the work should be carried out as two streams of PRs: 
 - one w.r.t. to the long-running branch for the OTS refactoring proper 
 - and the other w.r.t. main.

The main-side stream is to contain refactorings that carry majority of algorithmic/architectural changes that are necessary to make OTS-side refactorings straightforward, ideally being simple code movement. 
There are specific comments in the review illustrating this matter.

I believe work along these lines is worth doing as part of concluding this PR, but if this proposal feels unclear or overly complicated, let's discuss after you have evaluated the feedback.

File: lang/src/org/partiql/lang/ots_work/plugins/standard/types/DecimalType.kt [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
 import org.partiql.lang.eval.ExprValueType
-import org.partiql.lang.ots.interfaces.ScalarType
-import org.partiql.lang.ots.interfaces.TypeParameters
+import org.partiql.lang.ots_work.interfaces.CompileTimeType
+import org.partiql.lang.ots_work.interfaces.TypeParameters
+import org.partiql.lang.ots_work.interfaces.type.ScalarType
 import java.math.BigDecimal
 
 object DecimalType : ScalarType {
+    val compileTimeType: CompileTimeType = CompileTimeType(this, listOf(null, 0))
```
[the remark this answers is not in the document]
Comment (lziq) [reply lang/src/org/partiql/lang/ots_work/plugins/standard/types/DecimalType.kt:11]: This field is not a common field defined by `ScalarType` interface. It is just a helper field for the use in the standard package.

File: lang/src/org/partiql/lang/ast/passes/inference/StaticTypeExtensions.kt [hunk 2]
```diff
@@ -26,30 +30,19 @@ import org.partiql.lang.types.StaticScalarType
...
-private fun DecimalType.maxDigits(parameters: TypeParameters): Int = (parameters[0] ?: Int.MAX_VALUE) - parameters[1]!!
-
 /**
  * Casts [this] static to the given target type.
  *
  * This replicates the behavior of its runtime equivalent [ExprValue.cast].
  * @see [ExprValue.cast] for documentation.
  */
-internal fun StaticType.cast(targetType: StaticType): StaticType {
+internal fun StaticType.cast(targetType: StaticType, scalarTypeSystem: ScalarTypeSystem): StaticType {
```
[the remark this answers is not in the document]
Comment (lziq) [reply lang/src/org/partiql/lang/ast/passes/inference/StaticTypeExtensions.kt:45]: This file covers all the helper methods/classes used by `StaticTypeInferenceVisitorTransform.kt`. So I think it belongs to type inferencer.

File: lang/src/org/partiql/lang/ots_work/stscore/ScalarTypeSystem.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        ScalarOpId.BinaryConcat -> plugin.binaryConcatOp
+        ScalarOpId.Not -> plugin.notOp
+        ScalarOpId.Like -> plugin.likeOp
+    }
+
+    internal fun inferReturnType(scalarOpId: ScalarOpId, vararg argsType: CompileTimeType) =
+        inferReturnType(scalarOpId, argsType.toList())
+
+    internal fun inferReturnType(scalarOpId: ScalarOpId, argsType: List<CompileTimeType>) =
+        when (val scalarOp = getScalarOp(scalarOpId)) {
```
[the remark this answers is not in the document]
Comment (lziq) [reply lang/src/org/partiql/lang/ots_work/stscore/ScalarTypeSystem.kt:51]: Actually, I think we need this method in `ScalarTypeSystem`. It works as API between the plugin and the type inferencer. From type inferencer's side, it is something like, I can tell you which operator it is and what argument types we have, and please give me the correct return types. This whole method works as a black box for type inferencer to use and figures out which scalar operator to call from plugin.

[the remark this answers is not in the document]
Comment (lziq) [reply lang/src/org/partiql/lang/ots_work/stscore/ScalarTypeSystem.kt:51]: Even though now it looks a bit redundant, but later when we start to support operator/function overloading, there will be some code which checks all the available operators/functions, and find the most proper one to call. That part of code should be put here as well.

[the remark this answers is not in the document]
Comment (vgapeyev) [reply lang/src/org/partiql/lang/ots_work/stscore/ScalarTypeSystem.kt]: There will be a more detailed explanation why this is problematic, 
in lang/src/org/partiql/lang/util/ots_work/PluginExtensions.kt

**Review by vgapeyev — APPROVED** [review]
I have a couple fairly substantial new comments, which I think should be addressed. 
However, at this point, I think it is better to merge this PR into the base feature branch and do a fresh PR or a couple from there.

**Comment by lziq** [comment]
Followup on this PR, https://github.com/partiql/partiql-lang-kotlin/pull/774

**Pull request merged**